common-libs/internal-api-clients/src/helpers/client.ts

```diff
@@ -0,0 +1,149 @@
+import keyBy from 'lodash.keyby'
+import mapValues from 'lodash.mapvalues'
+import FetchType from 'node-fetch'
+import { SdkError } from '@affinidi/tools-common'
+
+import { GenericApiSpec } from '../types/openapi'
+import { ParseSpec } from '../types/openapiParser'
+import { ResponseForOperation, RequestOptionsForOperation, RequestOptions } from '../types/request'
+import { BuildApiTypeWithoutConstraint, BuiltApiOperationType, BuiltApiType } from '../types/typeBuilder'
+import { createAdditionalHeaders, createHeaders } from './headers'
+
+let fetch: typeof FetchType
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+if (!fetch) {
+  fetch = require('node-fetch')
+}
+
+export type ThisData = {
+  accessApiKey: string
+  sdkVersion?: string
+  serviceUrl: string
+}
+
+type MethodTypeByOperation<TOperation extends BuiltApiOperationType> = (
+  this: ThisData,
+  requestOptions: RequestOptionsForOperation<TOperation>,
+) => Promise<ResponseForOperation<TOperation>>
+
+type ClientTypeByApi<TApi extends BuiltApiType> = {
+  [key in keyof TApi]: MethodTypeByOperation<TApi[key]>
+}
+
+export type ClientTypeByRawSpec<TRawSpec extends GenericApiSpec> = ClientTypeByApi<
+  BuildApiTypeWithoutConstraint<ParseSpec<TRawSpec>>
+>
+
+export type ClientOptions = Omit<ThisData, 'serviceUrl'>
+
+export type ClientFactoryByRawSpec = {
+  <TRawSpec extends GenericApiSpec>(rawSpec: TRawSpec): ClientTypeByRawSpec<TRawSpec>
+}
+
+type GetRequestOptions<TOperation extends MethodTypeByOperation<any>> = Parameters<TOperation>[0]
+export type GetParams<TOperation extends MethodTypeByOperation<any>> = GetRequestOptions<TOperation>['params']
+
+const parseSpec = (rawSpec: GenericApiSpec) => {
+  const basePath = rawSpec.servers[0].url
+
+  const spec = Object.entries(rawSpec.paths).flatMap(([operationPath, operation]) => {
+    const path = `${basePath}${operationPath}`
+
+    return (['get', 'post', 'put', 'delete'] as const).flatMap((method) => {
```

Comment:
This hardcoded list worries me. If we're going to add some PATCH routes on our API, it won't work here until we update this list?

Reply:
If we're going to add some PATCH routes, build will fail until we add PATCH support because of https://github.com/affinityproject/affinidi-core-sdk/blob/f9b09b397570a903567e0d378c3dc5bb7d5f4db9/common-libs/internal-api-clients/src/types/openapi.ts#L136

(Also this code wasn't changed, before this PR it just resided in another file)

But I can add support for other methods if you feel like we need it here.